AwsCryptographicMaterialProviders/dafny/AwsCryptographicMaterialProviders/src/CMMs/CachingCMM.dfy

```diff
@@ -0,0 +1,1100 @@
+                 && input == Seq.Last(underlyingCMM.History.GetEncryptionMaterials).input
+                 && output == Seq.Last(underlyingCMM.History.GetEncryptionMaterials).output
+
+                 //= aws-encryption-sdk-specification/framework/caching-cmm.md#get-encryption-materials
+                 //= type=implication
+                 //# If the [algorithm suite](algorithm-suites.md) on the returned encryption material
+                 //# does not contain an [Identity KDF](algorithm-suites.md#identity-kdf),
+                 //# and the newly constructed [cache entry is within limits](#cache-entry-within-limits),
+                 //# the caching CMM MUST add the encryption materials obtained from the underlying CMM into the underlying CMC.
+                 && var encryptionMaterials := output.value.encryptionMaterials;
+                 (
+                   && !encryptionMaterials.algorithmSuite.kdf.IDENTITY?
+                   && (exists now
+                         :: cacheEntryWithinLimits(
+                              creationTime := now,
+                              bytesUsed    := input.maxPlaintextLength.value,
+                              messagesUsed := 1,
+                              now          := now
+                            ))
+                   ==>
+                     && |cache.History.PutCacheEntry| == |old(cache.History.PutCacheEntry)| + 1
+                     && var PutCacheEntryInput := Seq.Last(cache.History.PutCacheEntry).input;
+                     && PutCacheEntryInput.materials == Types.Encryption(output.value.encryptionMaterials)
+                     && PutCacheEntryInput.identifier == Seq.Last(cache.History.GetCacheEntry).input.identifier
+                     && PutCacheEntryInput.bytesUsed == input.maxPlaintextLength
+                 )
+                 //= aws-encryption-sdk-specification/framework/caching-cmm.md#get-encryption-materials
+                 //= type=implication
+                 //# If the [algorithm suite](algorithm-suites.md) on the returned encryption material
+                 //# contains an an [Identity KDF](algorithm-suites.md#identity-kdf),
+                 //# or the newly constructed [cache entry is not within limits](#cache-entry-within-limits),
+                 //# the caching CMM MUST NOT store the encryption materials in the underlying CMC.
+                 && (
+                   || encryptionMaterials.algorithmSuite.kdf.IDENTITY?
+                   || (exists now
+                         :: !cacheEntryWithinLimits(
+                              creationTime := now,
+                              bytesUsed    := input.maxPlaintextLength.value,
+                              messagesUsed := 1,
+                              now          := now
+                            ))
+                   ==>
+                     && |cache.History.PutCacheEntry| == |old(cache.History.PutCacheEntry)|
+                 )
```

Comment:
Note: I've always wondered about how KDF is used for the Caching CMM.
What I don't understand is when the KDF is invoked.
i.e: We have returned and cached materials b/c a KDF is in the Algorithm Suite.
But when are we are going to KDF?

Only during content encryption?

Reply:
Yes, during content encryption.
The materials are not "smart" so they leave the do the KDF to the caller.